lms/models/oauth2_token.py

```diff
@@ -0,0 +1,60 @@
+"""Database model for persisting OAuth 2 tokens."""
+import datetime
+
+import sqlalchemy as sa
+
+from lms.db import BASE
+
+
+class OAuth2Token(BASE):
+    """An OAuth 2.0 access token, refresh token and expiry time."""
+
+    __tablename__ = "oauth2_token"
+    __table_args__ = (sa.UniqueConstraint("user_id", "consumer_key"),)
+
+    id = sa.Column(sa.Integer(), autoincrement=True, primary_key=True)
+
+    #: The LTI user_id of the LMS user who this access token belongs to.
+    user_id = sa.Column(sa.UnicodeText(), nullable=False)
+
+    #: The LTI consumer_key (oauth_consumer_key) of the application instance
+    #: that this access token belongs to.
+    consumer_key = sa.Column(
+        sa.String(),
+        sa.ForeignKey("application_instances.consumer_key", ondelete="cascade"),
+        nullable=False,
+    )
+
+    #: The ApplicationInstance that this access token belongs to.
+    application_instance = sa.orm.relationship(
+        "ApplicationInstance", back_populates="access_tokens"
+    )
+
+    #: The OAuth 2.0 access token, as received from the authorization server.
+    access_token = sa.Column(sa.UnicodeText(), nullable=False)
+
+    #: The OAuth 2.0 refresh token, as received from the authorization server.
+    #: This will be null if the authorization server didn't provide a refresh
+    #: token.
+    refresh_token = sa.Column(sa.UnicodeText())
+
+    #: The lifetime in seconds of the access token, as given to us by the
+    #: authorization server.
+    #: Null if the authorization server didn't provide an expires_in number.
+    expires_in = sa.Column(sa.Integer())
+
+    #: The time when this access token was created.
+    created = sa.Column(
```

Comment:
Will this get updated when the access token is refreshed? If not, then `expires_in` isn't useful because it is a relative value that needs to be combined with `created` to get an absolute expiry time for comparison against the current time. If it does, then the documentation should make this clear.

Reply:
It'll get updated. This refers to when the access token was created (the `access_token` and `expires_in` fields get changed on refresh) rather than to when the DB row itself got created (which I consider an internal detail). To put it another way: refreshing could work by deleting any existing matching row and then inserting a new row, or it could work by updating an existing row in place, which way it works is an internal implementation detail.

Reply:
I've pushed a commit to rename this to `received_at` and add comments: https://github.com/hypothesis/lms/pull/611/commits/8874915aeb06990440d47a9199779a334ccc4c8e I've already merged the DB migration but haven't run it, so I snuck a change to the migration into this commit too